File: packages/ts/src/events/media.events.ts
/**
 * Media service event types for Kafka / NATS consumers.
 * Any NestJS service that subscribes to media events should import from here.
 */

export enum MediaEventType {
  UploadCreated       = 'media.upload.created',
  UploadCompleted     = 'media.upload.completed',
  UploadAborted       = 'media.upload.aborted',

  ScanStarted         = 'media.scan.started',
  ScanCompleted       = 'media.scan.completed',
  ScanFailed          = 'media.scan.failed',

  ProcessingStarted   = 'media.processing.started',
  ProcessingCompleted = 'media.processing.completed',
  ProcessingFailed    = 'media.processing.failed',

  MediaReady          = 'media.ready',

  DeleteRequested     = 'media.delete.requested',
  MediaDeleted        = 'media.deleted',
  DeleteFailed        = 'media.delete.failed',
}

export interface MediaVersionInfo {
  versionType: string;
  objectKey: string;
  contentType: string;
  size: number;
  width?: number;
  height?: number;
  durationMs?: number;
}

export interface MediaUploadCreatedPayload {
  mediaId: string;
  userId: string;
  deliveryId?: string;
  fileName: string;
  contentType: string;
  mediaType: string;
  size: number;
  uploadId: string;
}

export interface MediaUploadCompletedPayload {
  mediaId: string;
  userId: string;
  objectKey: string;
  size: number;
  contentType: string;
  mediaType: string;
}

export interface MediaReadyPayload {
  mediaId: string;
  userId: string;
  mediaType: string;
  contentType: string;
  versions: MediaVersionInfo[];
}

export interface MediaDeletedPayload {
  mediaId: string;
  userId: string;
  at: string; // ISO 8601
}

export interface MediaScanCompletedPayload {
  mediaId: string;
  userId: string;
  infected: boolean;
  threat?: string;
}

/**
 * Generic Kafka event envelope — matches the Go EventEnvelope struct.
 * All media events are wrapped in this structure.
 */
export interface MediaEventEnvelope<T = unknown> {
  eventId: string;
  eventType: MediaEventType;
  traceId?: string;
  timestamp: number; // unix milliseconds
  payload: T;
}

File: packages/go/events/media_events.go
package events

import (
	"encoding/json"
	"fmt"
	"time"
)

// MediaEventType is the canonical event type identifier for Kafka messages.
type MediaEventType string

const (
	MediaUploadCreated       MediaEventType = "media.upload.created"
	MediaUploadCompleted     MediaEventType = "media.upload.completed"
	MediaUploadAborted       MediaEventType = "media.upload.aborted"
	MediaScanStarted         MediaEventType = "media.scan.started"
	MediaScanCompleted       MediaEventType = "media.scan.completed"
	MediaScanFailed          MediaEventType = "media.scan.failed"
	MediaProcessingStarted   MediaEventType = "media.processing.started"
	MediaProcessingCompleted MediaEventType = "media.processing.completed"
	MediaProcessingFailed    MediaEventType = "media.processing.failed"
	MediaReady               MediaEventType = "media.ready"
	MediaDeleteRequested     MediaEventType = "media.delete.requested"
	MediaDeleted             MediaEventType = "media.deleted"
	MediaDeleteFailed        MediaEventType = "media.delete.failed"
)

// MediaUploadCreatedPayload is emitted when a new upload session is created.
type MediaUploadCreatedPayload struct {
	MediaID     string `json:"mediaId"`
	UserID      string `json:"userId"`
	DeliveryID  string `json:"deliveryId,omitempty"`
	FileName    string `json:"fileName"`
	ContentType string `json:"contentType"`
	MediaType   string `json:"mediaType"`
	Size        int64  `json:"size"`
	UploadID    string `json:"uploadId"`
}

// MediaUploadCompletedPayload is emitted when a multipart upload is finalised.
type MediaUploadCompletedPayload struct {
	MediaID     string `json:"mediaId"`
	UserID      string `json:"userId"`
	FileName    string `json:"fileName"`
	ObjectKey   string `json:"objectKey"`
	Size        int64  `json:"size"`
	ContentType string `json:"contentType"`
	MediaType   string `json:"mediaType"`
}

// MediaReadyPayload is emitted when all processing is complete and the media is available.
type MediaReadyPayload struct {
	MediaID     string             `json:"mediaId"`
	UserID      string             `json:"userId"`
	FileName    string             `json:"fileName"`
	Size        int64              `json:"size"`
	MediaType   string             `json:"mediaType"`
	ContentType string             `json:"contentType"`
	Versions    []MediaVersionInfo `json:"versions"`
}

// MediaVersionInfo describes one processed rendition.
type MediaVersionInfo struct {
	VersionType string `json:"versionType"`
	ObjectKey   string `json:"objectKey"`
	ContentType string `json:"contentType"`
	Size        int64  `json:"size"`
	Width       int32  `json:"width,omitempty"`
	Height      int32  `json:"height,omitempty"`
	DurationMS  int64  `json:"durationMs,omitempty"`
}

// MediaDeletedPayload is emitted when a media item is fully deleted from S3.
type MediaDeletedPayload struct {
	MediaID string    `json:"mediaId"`
	UserID  string    `json:"userId"`
	At      time.Time `json:"at"`
}

// MediaScanCompletedPayload is emitted after antivirus scanning.
type MediaScanCompletedPayload struct {
	MediaID  string `json:"mediaId"`
	UserID   string `json:"userId"`
	Infected bool   `json:"infected"`
	Threat   string `json:"threat,omitempty"`
}

// EventEnvelope is the canonical Kafka/NATS event envelope.
// All durable events should preserve correlation and causation information.
type EventEnvelope struct {
	EventID       string          `json:"eventId"`
	EventType     string          `json:"eventType"`      // e.g. "payment.authorized", "media.ready"
	EventVersion  int             `json:"eventVersion"`   // schema version
	OccurredAt    int64           `json:"occurredAt"`     // unix milliseconds
	Producer      string          `json:"producer"`       // service name that produced the event
	AggregateType string          `json:"aggregateType"`  // e.g. "payment", "media"
	AggregateID   string          `json:"aggregateId"`    // ID of the aggregate
	CorrelationID string          `json:"correlationId"`  // ID linking related events across services
	CausationID   string          `json:"causationId"`    // ID of the command that caused this event
	Payload       json.RawMessage `json:"payload"`
}

// NewEventEnvelope creates a new EventEnvelope.
func NewEventEnvelope(eventID string, eventType string, traceID string, payload interface{}) (*EventEnvelope, error) {
	payloadBytes, err := json.Marshal(payload)
	if err != nil {
		return nil, fmt.Errorf("marshal event payload: %w", err)
	}
	return &EventEnvelope{
		EventID:       eventID,
		EventType:     eventType,
		EventVersion:  1,
		OccurredAt:    time.Now().UnixMilli(),
		Producer:      "unknown",
		Payload:       payloadBytes,
	}, nil
}

// NewMediaEventEnvelope is a backward-compatible wrapper for media events.
func NewMediaEventEnvelope(eventID string, eventType MediaEventType, traceID string, payload interface{}) (*EventEnvelope, error) {
	return NewEventEnvelope(eventID, string(eventType), traceID, payload)
}

// Accepts any string event type for cross-domain use.
func MarshalEnvelope(eventID string, eventType string, traceID string, payload interface{}) ([]byte, error) {
	env, err := NewEventEnvelope(eventID, eventType, traceID, payload)
	if err != nil {
		return nil, err
	}
	return json.Marshal(env)
}

// MarshalMediaEnvelope is a backward-compatible wrapper for media events.
func MarshalMediaEnvelope(eventID string, eventType MediaEventType, traceID string, payload interface{}) ([]byte, error) {
	return MarshalEnvelope(eventID, string(eventType), traceID, payload)
}

// UnmarshalEnvelope parses a raw Kafka/NATS message into an EventEnvelope.
func UnmarshalEnvelope(data []byte) (*EventEnvelope, error) {
	var env EventEnvelope
	if err := json.Unmarshal(data, &env); err != nil {
		return nil, fmt.Errorf("unmarshal event envelope: %w", err)
	}
	return &env, nil
}